dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Flash/Pipeline/Schedule/TaskQueues/FIFOQueryIdCache.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/TaskQueue.h>
+#include <Flash/ResourceControl/LocalAdmissionController.h>
+
+#include <mutex>
+#include <unordered_map>
+
+namespace DB
+{
+template <typename NestedQueueType>
+class ResourceControlQueue
+    : public TaskQueue
+    , private boost::noncopyable
+{
+public:
+    ResourceControlQueue() = default;
+    ~ResourceControlQueue() override = default;
+
+    void submit(TaskPtr && task) override;
+
+    void submit(std::vector<TaskPtr> & tasks) override;
+
+    bool take(TaskPtr & task) override;
+
+    void updateStatistics(const TaskPtr & task, ExecTaskStatus, size_t inc_value) override;
+
+    bool empty() const override;
+
+    void finish() override;
+
+    void cancel(const String & query_id, const String & resource_group_name) override;
+
+#ifndef DBMS_PUBLIC_GTEST
+private:
+#endif
+    // <resource_group_name, resource_group_task_queues>
+    using ResourceGroupTaskQueue = std::unordered_map<std::string, std::shared_ptr<NestedQueueType>>;
+
+    // <priority, corresponding_task_queue, resource_group_name>
+    using ResourceGroupInfo = std::tuple<UInt64, std::shared_ptr<NestedQueueType>, std::string>;
+    using CompatorType = bool (*)(const ResourceGroupInfo &, const ResourceGroupInfo &);
+    using ResourceGroupInfoQueue = std::priority_queue<ResourceGroupInfo, std::vector<ResourceGroupInfo>, CompatorType>;
+
+    // Index of ResourceGroupInfo.
+    static constexpr auto InfoIndexPriority = 0;
+    static constexpr auto InfoIndexPipelineTaskQueue = 1;
+    static constexpr auto InfoIndexResourceGroupName = 2;
+    static constexpr auto DEFAULT_WAIT_INTERVAL_WHEN_RUN_OUT_OF_RU = std::chrono::seconds(1);
```

Comment:
Why is this needed?

Reply:
When all RG runs out of RU, wait for `DEFAULT_WAIT_INTERVAL_WHEN_RUN_OUT_OF_RU` to let LAC fetch tokens from GAC.

Reply:
How about proactively notifying `ResourceGroupQueue` after `LAC` gets the token?

Reply:
If ResourceGroupQueue waits exactly at 4.9s, then ResourceGroupQueue will wait 0.9s longer.
LAC   ----------
RCQ              -------

Reply:
Adding a callback to LAC, in order to notify RCQ after fetching the token?
```
LAC.register([&]() {
    condition_cv.notify();
})
```

Reply:
changed